doc/showinfilemanager.1.md
---
title: SHOW IN FILE MANAGER
section: 1
header: General Commands Manual
footer: showinfilemanager
date: 2021-08-13
---
# NAME
showinfilemanager - open the file manager and select files in it

# SYNOPSIS
**showinfilemanager** [**-h**] [**--version**] [**--verbose**] [**--debug**] [**URIs ...**]

# DESCRIPTION
**showinfilemanager** opens the system file manager and optionally selects files in it. The point is not to *open* the 
files, but to *select* them in the file manager, thereby highlighting the files and allowing the user to quickly do
something with them. 

For file managers that do not support specifying files to select, this program will instruct them to display the
directory the file is found in. 

**URIs**
: URIs or paths of files or directories.

**-h**, **--help** 
: Display help message.

**--version** 
: Show program's version number and exit.

**-f**, **--file-manager** **FILE_MANAGER**
: File manager to run

**-s**, **--select-folder**
: Select folder instead of displaying its contents. By default, when a URI or path is a directory and
  not a file, the directory itself is shown in the file manager. This option changes the default and 
  selects the folder, displaying it in its parent directory.

**--verbose**
: Display command being run to stdout.

**--debug**
: Output debugging information to stdout.

# EXAMPLES
**showinfilemanager ~/myfile.txt**
: Open the system file manager and select *~/myfile.txt* in the user's home directory.

**showinfilemanager -f dolphin \*.txt**
: Open the dolphin file manager and select all the files with extension *txt* in the current directory.

**showinfilemanager D:\\Documents\\\*.docx**
: Open the system file manager and select all the Word documents in the Documents directory on the D: drive.

**showinfilemanager -s /Users/***
: Open the system file manager and select all the files and directories in the Users folder

**showinfilemanager file:///home/user/first%20file.txt file:///home/user/second%20file.txt**
: Open the system file manager and select 'first file.txt' and 'second file.txt' in the user's home directory.

**showinfilemanager /etc/hosts /home/user/.bashrc**
: Open the system file manager and select the hosts file in one file manager window, and the user's .bashrc file in 
another.


# AUTHORS
Damon Lynch  
damonlynch@gmail.com

# BUGS
Submit bug reports online at: <https://github.com/damonlynch/showinfilemanager/issues>

# SEE ALSO
Full documentation and sources at: <https://github.com/damonlynch/showinfilemanager>

CHANGELOG.md
Changelog for Show in File Manager
==================================

0.9.0 (2021-09-08)
------------------
 - Add option to specify file manager to use from command line
 - Support [Double Commander](https://doublecmd.sourceforge.io/)
 - Support [Krusader](https://krusader.org/)
 - Support [SpaceFM](https://ignorantguru.github.io/spacefm/)
 - Support [fman](https://fman.io/)

0.0.8 (2021-08-31)
------------------
 - Support [CutefishOS](https://en.cutefishos.com/).
 - Support [Index File Manager](https://invent.kde.org/maui/index-fm).

0.0.7 (2021-08-20)
------------------
 - Use `--select` command line switch in caja 1.26 or newer.

0.0.6 (2021-08-19)
------------------
 - Remove get_ prefix from package level function names.

0.0.5 (2021-08-19)
------------------
 - Add setup.py for man page generation.
 - Improve README to clarify installation and usage.
 - Parse filename globs passed via the command line on Windows.
 - Use win32 API to execute explorer.exe on Windows, allowing for
   multiple file selection.

0.0.4 (2021-08-14)
------------------
 - Update README to include installation instructions.
 - Include CHANGELOG.md in package.
 - Generate man page for use in Linux.
 - Improve command line argument documentation.

0.0.3 (2021-08-12)
------------------
 - Update README.

0.0.2 (2021-08-12)
------------------
 - Move config data from `__about__.py` to static config in 
   `pyproject.toml` [(PEP 621)](https://www.python.org/dev/peps/pep-0621/).
 - Added command line arguments `--debug` and `--verbose`.

0.0.1 (2021-08-12)
------------------
 - Initial release.

src/showinfilemanager/showinfilemanager.py
import os
import platform
import shlex
import subprocess
import sys
from typing import Optional, Tuple, Union, Sequence
import urllib.parse
import shutil
import argparse
import pathlib


import showinfilemanager.linux as linux
import showinfilemanager.__about__ as __about__
import showinfilemanager.__init__ as __init__
from showinfilemanager.constants import FileManagerType, Platform

_valid_file_manager_probed = False
_valid_file_manager = None
_valid_file_manager_type = None
_platform_probed = False
_platform = None



def get_stock_file_manager() -> str:
    """
    Get stock file manager for this operating system / desktop.

    Exceptions are not caught.

    :return: executable name
    """

    if _platform == Platform.windows:
        file_manager = 'explorer.exe'
    elif _platform == Platform.linux:
        file_manager = linux.get_stock_linux_file_manager()
    else:
        raise NotImplementedError

    assert shutil.which(file_manager) is not None
    return file_manager


def get_user_file_manager() -> str:
    """
    Get the file manager as set by the user.

    Exceptions are not caught.

    :return: executable name
    """

    if _platform == Platform.windows:
        file_manager = 'explorer.exe'
    elif _platform == Platform.linux:
        file_manager = linux.get_user_linux_file_manager()
    else:
        raise NotImplementedError

    assert shutil.which(file_manager) is not None
    return file_manager


def get_valid_file_manager() -> str:
    """
    Get user's file manager, falling back to using sensible defaults for the desktop / OS.

    All exceptions are caught, except those if this platform is not supported by this module.

    :return: If the user's default file manager is set and it is known by this module, then
    return it. Otherwise return the stock file manager, if it exists.
    """

    if _platform == Platform.windows:
        file_manager = 'explorer.exe'
    elif _platform == Platform.linux:
        file_manager = linux.get_valid_linux_file_manager()
    else:
        raise NotImplementedError
    return file_manager


def show_in_file_manager(path_or_uri: Optional[Union[str, Sequence[str]]] = None,
                         file_manager: Optional[str] = None) -> None:
    """
    Open the system file manager and display an optional directory, or items in the  directory.

    If there is no valid file manager found on the system, do nothing. A valid file manager
    is a file manager this module knows about.

    :param path_or_uri: zero or more files or directories to open, specified as a single URI
     or valid path, or a sequence of URIs/paths.
    :param file_manager: executable name to use. If not specified, then get_valid_file_manager()
     will determine which file manager to use.
    """

    global _valid_file_manager
    global _valid_file_manager_type

    if not file_manager:
        _set_valid_file_manager()
        file_manager = _valid_file_manager

    if not file_manager:
        return

    if not path_or_uri:
        arg = ''
        uris = ''
    else:
        if isinstance(path_or_uri, str):
            # turn the single path / URI into a Tuple
            path_or_uri = path_or_uri,

        uris = ''
        for pu in path_or_uri:
            # Were we passed a URI or simple path?
            parse_result = urllib.parse.urlparse(pu)
            if parse_result.scheme:
                uri = pu
            else:
                # Convert the path to a URI
                uri = pathlib.Path(os.path.abspath(pu)).as_uri()
                parse_result = None

            if _valid_file_manager_type == FileManagerType.dir_only_uri:
                # Show only the directory; do not attempt to select the file
                if parse_result is None:
                    parse_result = urllib.parse.urlparse(uri)
                uri = urllib.parse.urlunparse(parse_result._replace(path=os.path.dirname(parse_result.path)))

            uris = '{} {}'.format(uris, uri)

        arg = ''
        if _valid_file_manager_type == FileManagerType.win_select:
            arg = '/select,'  # no trailing space is necessary on Windows
        elif _valid_file_manager_type == FileManagerType.select:
            arg = '--select '  # trailing space is necessary for this and subsequent entries
        elif _valid_file_manager_type == FileManagerType.show_item:
            arg = '--show-item '
        elif _valid_file_manager_type == FileManagerType.show_items:
            arg = '--show-items '

    # Some file managers must be passed only one or zero paths / URIs
    if file_manager in ('explorer.exe', 'pcmanfm'):
        uris = uris.split() or ['']
    else:
        uris = [uris]

    for u in uris:
        cmd = '{} {}{}'.format(file_manager, arg, u)
        print("Executing", cmd)
        # Do not check _platform here, it makes no sense
        if platform.system() != "Windows":
            args = shlex.split(cmd)
        else:
            args = cmd
        subprocess.Popen(args)


def _set_valid_file_manager() -> None:
    """
    Set module level global variables to set a valid file manager for this user in this desktop environment.
    """

    global _valid_file_manager_probed
    global _valid_file_manager
    global _valid_file_manager_type

    if not _valid_file_manager_probed:
        fm = get_valid_file_manager()
        if fm:
            _valid_file_manager = fm
            if _platform == Platform.windows:
                _valid_file_manager_type = FileManagerType.win_select
            elif _platform == Platform.linux:
                _valid_file_manager_type = linux.get_linux_file_manager_type(fm)
            else:
                raise NotImplementedError
        _valid_file_manager_probed = True


def _detect_platform() -> None:
    """
    Set module level global variables to detect the platform being run in.
    """

    global _platform_probed
    global _platform

    if not _platform_probed:
        _platform_probed = True
        system = platform.system()
        if system == "Windows" or linux.detect_wsl():
            _platform = Platform.windows
        elif system == "Linux":
            _platform = Platform.linux
        elif system == "Darwin":
            _platform = Platform.macos
        else:
            raise NotImplementedError


class Diagnostic:

    def __init__(self) -> None:
        _detect_platform()
        try:
            self.stock_file_manager = get_stock_file_manager()
        except Exception as e:
            self.stock_file_manager = str(e)
        try:
            self.user_file_manager = get_user_file_manager()
        except Exception as e:
            self.user_file_manager = str(e)
        try:
            self.valid_file_manager = get_valid_file_manager()
        except Exception as e:
            self.valid_file_manager = str(e)

        if _platform == Platform.linux:
            try:
                self.desktop = linux.get_linux_desktop()
            except:
                self.desktop = linux.LinuxDesktop.unknown
        else:
            self.desktop = ''

    def __str__(self) -> str:
        desktop = "Linux Desktop: {}\n".format(self.desktop.name) if self.desktop else ""
        file_managers = "Stock: {}\nUser's choice: {}\nValid: {}".format(
                    self.stock_file_manager,
                    self.user_file_manager,
                    self.valid_file_manager
        )
        return desktop + file_managers


def parser_options(formatter_class=argparse.HelpFormatter):
    parser = argparse.ArgumentParser(
        prog=__about__.__title__, description=__about__.__summary__, formatter_class=formatter_class
    )

    parser.add_argument(
        '--version', action='version', version='%(prog)s {}'.format(__init__.__version__)
    )

    parser.add_argument('--verbose', action='store_true')

    parser.add_argument('path', nargs='*')

    return parser


def main() -> None:
    parser = parser_options()

    args = parser.parse_args()

    if args.verbose or __init__.__version__ < '0.1.0':
        print(Diagnostic())

    show_in_file_manager(path_or_uri=args.path)


